text/3466-manuallydrop-deref-move.md

```diff
@@ -0,0 +1,159 @@
+
+# Explanation
+
+In today's Rust, `Box<T>` has the unique capability that it is possible to move out of a dereference of it.
+
+```rust
+let b: Box<String> = Box::new("hello world".to_owned());
+let s: String = *b; // `b`'s backing allocation is dropped here
+```
+
+Partial moves are also permitted.
+
+```rust
+let s: String;
+{
+    let b: Box<(String, String)> = Box::new(("hello".to_owned(), "world".to_owned()));
+    s = b.1;
+    // `b.0` and `b`'s backing allocation dropped here
+}
+```
+
+This RFC extends this capability to `ManuallyDrop`.
+
+```rust
+use core::mem::ManuallyDrop;
+
+let m: ManuallyDrop<(String, String)> = ManuallyDrop::new(("hello".to_owned(), "world".to_owned()));
+drop(m.1); // `m.1` moved out of here
+// `m.0` is never dropped
+```
+
+In addition, partial moves out of a `ManuallyDrop<T>`'s contents are allowed even when there is a `Drop` impl for `T`.
+
+```rust
+struct Foo(String, String);
+
+impl Drop for Foo {
+    fn drop(&mut self) {
+        println!("down to the ground");
+    }
+}
+
+let m: ManuallyDrop<Foo> = ManuallyDrop::new(Foo("hello".to_owned(), "world".to_owned()));
+let s: String = m.1; // `m.1` moved out of here
+// `m.0` is never dropped, and nothing is printed.
+```
+
+The example from the motivation section would be rewritten as:
+
+```rust
+impl Foo {
+    fn into_bar(self) -> Bar {
+        let m = core::mem::ManuallyDrop::new(foo);
+        Bar(m.0)
+    }
+}
+```
+
+# Drawbacks
+
+- Adds more "magic" to the language.
+- This change would give safe Rust code a new capability (moving out of fields of `Drop`-implementing structs). It's theoretically possible that existing `unsafe` is relying on this capability not existing for soundness:
+
+```rust
+use core::hint::unreachable_unchecked;
+
+/// If the bomb is dropped while armed,
+/// it explodes and triggers undefined behavior.
+pub struct Bomb {
+    is_armed: bool,
+}
+
+impl Drop for Bomb {
+    fn drop(&mut self) {
+        if self.is_armed {
+            println!("💥 BOOM 💥");
+
+            // SAFETY: the only way for arbitrary safe code to obtain a `Bomb`
+            // is via `DefuseWrapper::new()`. Because `DefuseWrapper` implements `Drop`,
+            // it is impossible to move the `Bomb` out of it.
+            // `DefuseWrapper`'s destructor ensures that
+            // `is_armed` is set to `false` before the `Bomb` is dropped,
+            // so this branch is unreachable.
+            unsafe { unreachable_unchecked(); }
+        }
+    }
+}
+
+/// Disarms the bomb before dropping it.
+pub struct DefuseWrapper {
+    pub bomb: Bomb,
+}
+
+impl Drop for DefuseWrapper {
+    fn drop(&mut self) {
+        self.bomb.is_armed = false;
+    }
+}
+
+impl DefuseWrapper {
+    pub fn new() -> Self {
+        DefuseWrapper {
+            bomb: Bomb {
+                is_armed: true,
+            }
+        }
+    }
+}
+```
+
+If code like the above example actually exists in the ecosystem, this proposal would make it unsound.
```

Comment:
I actually do have an unpublished crate that relies on fields not being able to be moved out if I implement `Drop`. Basically `SomeStruct` has a field with type `FieldReader`. The latter type is a ZST that implements `Deref` based on the address of `&FieldReader` (in this case, it reads other fields from `SomeStruct`), that relies on the fact that if I make `SomeStruct` implement `Drop`, and construction of `FieldReader` unsafe, then all pointers to `FieldReader` are guaranteed to exist only within `SomeStruct`, since moving out of `SomeStruct` isn't allowed.

It would be nice if there is a way to spell "can't move out of any fields of this struct" that does not involve a no-op drop implementation, or maybe there is a way already and I just don't know about it.

Reply:
What I would expect/was expecting is that in  practice, this guarantee would be provided via privacy. Your `FieldReader` would be a private field with a public `get()`/`get_mut()` accessor.

(Also, your use-case might be impacted by subobject provenance discussions, you might want to share details at https://github.com/rust-lang/unsafe-code-guidelines/issues/256)

Reply:
What I'm hoping to achieve with `FieldReader` is that I want it to be a public field, while emulating getters through `Deref`, so using accessor methods defeats its point entirely.

Reply:
This conversation reminds me of the debate over the [soundness of `partial-borrow` vs `replace_with`](https://lib.rs/crates/partial-borrow#readme-compatibility---language-assumption-about-owned-values-from-references).  In `partial-borrow` I also (ab)use a language inflexibility to allow the use of fields rather than accessors.

(Providing fields rather than accessors is more than just a simple ergonomic improvement to elide some `()`.  There are important borrowck features that are available only via borrowck's special handling of fields.)

So it seems both @fee1-dead's crate and `partial-borrow` would benefit from some way to get access to those "field-specific" borrowck behaviours.  In the meantime this is another case where not only (i) adding a new language feature can make existing code unsound (ii) that existing code is not theoretical and cannot be readily expressed a different way.

Reply:
> (Providing fields rather than accessors is more than just a simple ergonomic improvement to elide some `()`. There are important borrowck features that are available only via borrowck's special handling of fields.)

Yes. I forgot to mention this in my comment, but this is one big reason I'm using a `FieldReader` design, to allow having mutable references for two fields.